apps/ejabberd/src/mongoose_api_backend.erl

```diff
@@ -0,0 +1,234 @@
+        {ok, Data} ->
+            handle_body(Method, Data, Req2, State);
+        {error, _Reason} ->
+            error_response(bad_request, Req2, State)
+    end.
+
+handle_body(Method, Data, Req, #backend_state{command_category = Category}=State) ->
+    Action = method_to_action(Method),
+    Commands = ?COMMANDS_ENGINE:list(admin, Category),
+    [Command] = [C || C <- Commands, ?COMMANDS_ENGINE:action(C) =:= Action],
+    case check_and_extract_args(?COMMANDS_ENGINE:args(Command), Data) of
+        {ok, Args} ->
+            Result = execute_command(Args, Command),
+            handle_result(Result, Req, State);
+        {error, Type} ->
+            error_response(Type, Req, State)
+    end.
+
+handle_result({ok, Result}, Serializer, Req, State) ->
+    serialize(Result, Serializer, Req, State);
+handle_result(Other, _Serializer, Req, State) ->
+    handle_result(Other, Req, State).
+
+handle_result({ok, _Res}, Req, State) ->
+    %% TODO When POST add resource created to headers
+    %% TODO return appropriate status code
+    {ok, Req2} = cowboy_req:reply(201, [{<<"location">>, <<"http://localhost:5288/api/user">>}], Req),
+    {halt, Req2, State};
+handle_result({error, Error}, Req, State) ->
+    error_response(Error, Req, State);
+handle_result(no_call, Req, State) ->
+    error_response(not_implemented, Req, State).
+
+serialize(Data, Serializer, Req, State) ->
+    {Serializer:serialize(Data), Req, State}.
+
+register_handler(Base, Command) ->
+    {[Base, create_url_path(Command)],
+      ?MODULE, [{command_category, ?COMMANDS_ENGINE:category(Command)}]}.
+
+get_allowed_methods() ->
+    Commands = ?COMMANDS_ENGINE:list(admin),
+    [action_to_method(?COMMANDS_ENGINE:action(Command)) || {_Name, Command} <- Commands].
+
+execute_command(Args, Command) ->
+    ?COMMANDS_ENGINE:execute(admin, ?COMMANDS_ENGINE:name(Command), Args).
+
+create_url_path(Command) ->
+    "/" ++ maybe_add_bindings(category_to_resource(?COMMANDS_ENGINE:category(Command)), Command).
+
+%% for now, might be GET of form http://api/users/:domain/:username
+%% instead of                    http://api/users/domain/:domain/username/:username
+maybe_add_bindings(Base, Command) ->
+    Action = ?COMMANDS_ENGINE:action(Command),
+    Args = ?COMMANDS_ENGINE:args(Command),
+    case Action of
+        read ->
+            add_bindings(Base, Args);
+        update ->
+            add_bindings(Base, Args);
+        delete ->
+            add_bindings(Base, Args);
+        _ ->
+            Base
+    end.
+
+add_bindings(Base, Args) ->
+    Suffix = ["/:" ++ atom_to_list(ArgName)  || {ArgName, _} <- Args],
+    lists:flatten(Base ++ Suffix).
+
+extract_bindings(Bindings) ->
+    [Bind || {_BindingName, Bind} <- Bindings].
+
+category_to_resource(Category) when is_atom(Category) ->
+    atom_to_list(Category);
+category_to_resource(Category) when is_list(Category) ->
+    Category.
+
+check_and_extract_args(CommandsArgList, RequestArgList) ->
```

Comment:
You seem to be checking if arguments comply with command args definition - this is also done by command registry, so it is redundant here (I can export the check from command_registry module if you need to do it here). Your check also imposes a requirements on command args definition (see our discussion about add_bindings). An idea: if http frontend requires that command args comply to a particular pattern (a flat list of named args, for instance), it could check it at startup and simply do not expose those which do not comply.